### PR TITLE
ZipArchive.CreateEntry: include file type in default ExternalAttributes.

### PR DESCRIPTION
**Pull request by tmds**
For Unix-type zip files, ExternalAttributes value matches 'struct stat' st_mode.
Besides permissions it should also include the file type.

### CODE REVIEW

**Comment by tmds** [comment]
Setting default values for ExternalAttributes was added in https://github.com/dotnet/runtime/pull/70735.
I noticed this is supposed to have the file type while investigating https://github.com/dotnet/runtime/issues/76006.

@derwasp @eerhardt @adamsitnik @danmoseley ptal.

**Comment by eerhardt** [comment]
The build looks broken.

**Comment by tmds** [comment]
> The build looks broken.

It's fixed now.

**Comment by derwasp** [comment]
Oh, I learned something today!
@tmds can you also mention what happens when the filetype is not set?

**Comment by tmds** [comment]
> @tmds can you also mention what happens when the filetype is not set?

.NET ignores the file type bits and only looks at the permission bits.
I haven't looked at what other software does.

**Comment by eerhardt** [comment]
@tmds  - do you think this fix needs to be included in .NET 7? Was there a regression here?

**Comment by tmds** [comment]
> @tmds - do you think this fix needs to be included in .NET 7? Was there a regression here?

I gave it a try, and it looks like `unzip` behaves similar to .NET: it ignores the file type bits.
I see no need to backport this.

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your contribution @tmds!

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 1]
```diff
@@ -457,7 +457,7 @@ internal struct ZipCentralDirectoryFileHeader
         public ushort FileCommentLength;
         public int DiskNumberStart;
         public ushort InternalFileAttributes;
-        public uint ExternalFileAttributes;
+        public int ExternalFileAttributes;
```
Comment (eerhardt) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:460]: Why the change from `uint` to `int` here?

In general, since this doesn't represent a "signed" number, it is better to use `uint` - just like all the other fields in this struct. This seems pretty inconsistent now.

Comment (tmds) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:460]: The `public ZipArchiveEntry.ExternalAttributes` is an `int`, and I changed this to get rid of the casts.
I can change it back (and add back the casts) if you prefer?

Comment (eerhardt) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:460]: It seems so inconsistent now with the rest of the code and feels like an unnecessary change.

if

```C#
internal const int DefaultFileExternalAttributes =
```

was declared as a `uint`, I think there would be a lot less casting happening in your new code.

**Review by eerhardt — APPROVED** [review]
LGTM.

Thanks, @tmds!

**Comment by danmoseley** [comment]
is this ready to merge then?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 2]
```diff
@@ -117,10 +117,9 @@ internal ZipArchiveEntry(ZipArchive archive, string entryName)
 
             _compressedSize = 0; // we don't know these yet
             _uncompressedSize = 0;
-            UnixFileMode defaultEntryPermissions = entryName.EndsWith(Path.DirectorySeparatorChar) || entryName.EndsWith(Path.AltDirectorySeparatorChar)
-                                        ? DefaultDirectoryEntryPermissions
-                                        : DefaultFileEntryPermissions;
-            _externalFileAttr = (uint)(defaultEntryPermissions) << 16;
+            _externalFileAttr = entryName.EndsWith(Path.DirectorySeparatorChar) || entryName.EndsWith(Path.AltDirectorySeparatorChar)
```
Comment (Jozkee) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:120]:
```suggestion
            _externalFileAttr = Path.EndsInDirectorySeparator(entryName)
```

**Review by jozkee — APPROVED** [review]
